Add ability to use functionality with NVRs and channels

flacjacket opened the pull request:
There are a lot of changes here. If these generally look ok, I'd be happy to split some of the changes up into separate PRs or commits to make the review and merging easier.

Allow the Amcrest API to be used with channels.  Mostly, this includes
adding channel parameters to a lot of the API, but only for those
functions that differ per-channel.  In the case of properties with
setters that need to be per-channel, break them out to instead be getter
and setter functions.

Update the interfaces to include type information that can be used for
type checking internally and for external users.

Along with this, do a large-ish update to Python 3-style syntax,
including using f-strings for formatting, adding type annotations to the
interface, and dropping some unnecessary code (passing "utf-8" to
decode, parameters to super, etc).

Also, perform a full `black` formatting.

tchellomello reviewed (changes requested):
@flacjacket thanks for your contribution. Could you please fix the lint issues so we can get this merged?

tchellomello reviewed (changes requested):
@flacjacket  Thanks for your contribution!!! 

Could you please address the lint issues?

flacjacket commented:
@tchellomello I'd be happy to! Do you know what are still problematic lines? Looking at it, I think there are a few instances where there is a difference between what `black` would do and what `flake8` wants because of a difference in line length (black uses line length 88 and the check is looking for line length 79). There are also a couple instances of "redefinition of unused 'snapshot' from line 35", which is the check not accounting for the mypy overload definitions correctly. Do you know if there are any cases that I'm missing?

tchellomello reviewed (approved):
LGTM @flacjacket 

@dougsland @pnbruckner  can we have your input so we can merge it?

dougsland commented on src/amcrest/special.py:
```diff
@@ -30,80 +35,50 @@ def realtime_stream(self, channel=1, typeno=0, path_file=None):
 
         return ret.raw
 
-    def rtsp_url(self, channelno=None, typeno=None):
```
renaming function would break the API for users.

dougsland reviewed (commented):
Thanks for your patch. This is a mega large patch. If possible please break it in small patches for helping us to review.

flacjacket commented:
Thanks for the feedback! Sorry this took so long to come back to. I split this up across a few PRs, you can see #173 #174 #175 and #176, each one has the relevant changes in a single diff, but they are stacked on top of each other so you'll see all the changes in the PR view, but you can see the proposed change by just looking at the latest diff in the PR. I also changed this functionality to better preserve the current functionality and add the functionality to read various channels through new methods. Changes that are needed are noted in the PRs.

flacjacket commented on src/amcrest/audio.py:
```diff
@@ -78,61 +75,69 @@ def audio_send_stream(self, httptype=None,
 
-    @property
-    def audio_enabled(self):
+    def is_audio_enabled(self, *, channel: int = 0) -> bool:
```
The basis of this change is that to pull the property from a specified channel requires passing the `channel` argument, which cannot be done with a property. In the new PRs, the current `@property`'s are kept as-is, and the new methods are added to query other channels.

flacjacket commented on src/amcrest/__init__.py:
```diff
@@ -10,25 +8,68 @@
             port=port,
             user=user,
             password=password,
-            verbose=verbose,
```
I noticed that this wasn't used anywhere. I'm not sure what the deprecation policy is, so I removed this change from the other PR's, but other than breaking clients that pass in `verbose` should have no effect.

tchellomello reviewed (approved):
@dougsland @pnbruckner  Looks good to me guys after rebase it

flacjacket commented:
Thanks, this should be covered by the other PRs that were merged!
